Add plugin metadata support

grantila opened the pull request:
## Plugin metadata

This PR adds plugin metadata to plugins, in a field called `info`. It allows implementors (and plugin developers) to decorate plugins with:
 * Information about the plugin itself (coming from `package.json`, or totally custom)
 * "Ownership" of a plugin

The metadata (`PluginInfo`) contains name, version, owner, links etc. It will be deduced from `package.json`, and the way to do the decoration is for a plugin to:

```ts
import packageJson from '../package.json';

// ...

export const myPlugin = createPlugin({
  id: 'my-plugin',
  apis: [ /* ... */ ],
  routes: { /* ... */ },
  info: { packageJson }, // <-- add fields here, or just packageJson content
});
```

This means we can use the package.json as the basis for plugin metadata, and plugins can get information from the `package.json` about themselves or other plugins. The idea here is to re-use the `backstage` field for `backstage.name` and `backstage.role` etc.

An implementor can provide a callback `pluginInfoDecorator` to `createApp()` which will be called for each plugin, to allow further decorations of the plugin (extend with links to slack channels or whatnot).

You could argue that adding the package.json to each plugin will grow the webpack bundles. That's true, but it shouldn't be that bad though. A 1.5kB package.json is gzipped down to <600b. Having this would allow other future plugins to inspect this at runtime and indicate package dependency mismatches etc. Could be useful for more than just the metadata explicitly used here.

### Ownership

Alongside the metadata is ownership. It's very useful for implementors to be able to decorate their own _internal_ plugins with owners; what team/group owns a certain plugin. Combined with metadata, this would allow for error boundaries, help/support buttons etc to use that information to provide the user with relevant links to contact the corresponding owner for help / bug reports etc. Implementors can choose their own ownership model (maybe a field in package.json) and use the `pluginInfoDecorator` to decorate that to the `PluginInfo` blob.

This isn't only useful for _internal_ plugins though, but also for those plugins pulled in from this very open source repo (or any other). There's likely several teams being responsible for certain parts of a backstage implementation. You might have a core team being overall responsible for the implementation, maybe a documentation team, CI/CD team etc. This is where the `pluginOwners` option to `createApp()` comes in handy;

An implementor can create a mapping (e.g. in a yaml-file) between package names (npm package names) and internal team names, and these should (preferably) be **catalog groups**, or an entity-ref. With wild-card matching, you could create a mapping like this:

```yaml
# plugin-metadata.yaml
owners:
  - '*techdocs*': the-doc-team
  - '@backstage/plugin-ci*': cicd-team
  - '@backstage/plugin-jenkins*': cicd-team
  - '*': the-core-team # Fallback unless specified above
```

You'd then:

```ts
import pluginMetadata from './plugin-metadata.yaml';

createApp({
  pluginOwners: pluginMetadata.owners,
  // ...
});
```

A follow-up PR will use this.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on packages/core-plugin-api/src/plugin/types.ts:
```diff
@@ -74,6 +75,62 @@ export type PluginFeatureFlagConfig = {
+  /**
+   * Plugin name; defaults to `backstage.name` in package.json
+   */
+  name?: string;
```
There is already a plugin ID, so feeling this would be unfortunate. Perhaps a lot of this can be bundled up in some kind of "package metadata" instead? I think overall it would perhaps be good to group these properties up a bit, perhaps one metadata chunk, and another ownership & support chunk. Decoration form the app would then only be done of ownership & support

grantila replied:
Would `title` be better?

grantila replied:
Resolved by removing `name`

Rugvip commented on packages/core-plugin-api/src/plugin/types.ts:
```diff
@@ -88,6 +145,7 @@ export type PluginConfig<
   routes?: Routes;
   externalRoutes?: ExternalRoutes;
   featureFlags?: PluginFeatureFlagConfig[];
+  info?: PluginConfigInfo;
```
This might seem a bit silly, but I'm thinking it's worth making the `package.json`s lazy-loaded or in some other way make sure we're not eagerly loading in too much data. Considering for example the `package.json` of the catalog, it's roughly 2000 bytes. Make that 100 plugins and you've got 200kB added to the main bundle.

grantila replied:
Yeah, I just `cat`ed **all** the package.jsons in plugins/ and packages/ now, it's ~250kB although gzipped down to 17kB. Which is really not that much, considering..

Lazy-loading is an interesting idea, but it makes other things complicated, like, if you want to use this data in code, to e.g. log (as with #10259) or if you want to view all plugins (#10260) or if you want ErrorBoundaries, Analytics and other things to take advantage of this metadata, like ownership (based on package.json `name`), all of this would have to be async.

I'm really not sure what'd be the right procedure. Maybe a webpack filter for `package.json` imports that cuts off all unneccesary things (*dependencies) or only filters what we need. That'd trim the final size. Perhaps something like an import loader a la `'plugin-metadata!../package.json'`

grantila replied:
resolved by allowing lazy loaded (dynamically imported) `package.json`

Rugvip commented on packages/core-app-api/src/app/PluginDecoration.ts:
```diff
@@ -0,0 +1,141 @@
+    return match?.[1];
+  };
+
+  public decoratePlugin = <Plugin extends CompatiblePlugin>(plugin: Plugin) => {
```
Tempted to structure the plugin info API layer in a way that we don't need this direct mutation here, but I think we can leave that a bit and focus on the other bits first

grantila replied:
Yeah I didn't really know what to do tbh regarding this compatible-thingie (remove the compatibility altogether after 1.0?)

Rugvip reviewed (commented):
Loving the idea of an ownership mapping that can be applied both bottom-up from the plugin or pushed down through the app. Feeling that's a huge leap towards getting proper ownership (and the support button) in place. 👍 🚀

grantila commented:
Would be happy to do some renaming too, "decoration"/"info" -> "metadata" e.g. if that makes more sense. And add tests for it. But it makes more sense to nail the design/architecture of it, e.g. whether to go with lazy-loading.

grantila commented:
Changed to lazy-loaded `package.json` content for per-plugin metadata and refactored quite a bit.

 * (in core-app-api) Aside `AppManager` is now a class `PluginMetadataExtender` which is used for app-level metadata decorations (mapping corporate _owners_ to plugins by name) and allowing any custom metadata decoration.
 * (in core-plugin-api) Aside `Plugin` is now a separate file with `getPluginMetadata()` which performs the metadata extraction asynchronously. It allows lazy-loaded `package.json`s, and extracts fields from it. It then hands over to the app-level metadata decorations (for e.g. ownership mapping).

When creating a plugin, you can simply let `info` be a dyn-loaded import:

```ts
export const myPlugin = createPlugin({
  id: 'my-plugin',
  info: () => import('../package.json'),
  /* ... */
});
```

But you can also specify certain fields if you want, and let only the packageJson part be lazy-loaded:

```ts
export const myPlugin = createPlugin({
  id: 'my-plugin',
  info: {
    packageJson: () => import('../package.json'),
    description: 'My plugin',
    links: [{ title: 'Somewhere', url: 'https://somewhere.com' }],
  },
  /* ... */
});
```

On a plugin object (`BackstagePlugin`) is now `getInfo()` which asynchronously returns a metadata object `Promise<PluginInfo>` given the above procedure.

freben commented:
@grantila Ping on this one - DCO and build warnings

freben commented:
damn it, misclick

Rugvip commented:
For when you're back: we'll want to figure out how this fits into #11404 too. I'm feeling at least the context has some significant overlap, but perhaps the method of configuration too?

grantila commented:
@Rugvip this is up-to-date with master now, with DCO, and `name` is removed. I think this is pretty clean now, please take a look.

I could maybe add some documentation for this, for:
 * How to setup the app (and configure app-level ownership mapping)
 * How to add metadata to a plugin (for `createPlugin`)

Where, if so, should I add that?

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -220,6 +220,8 @@ export type BackstagePlugin<
   getApis(): Iterable<AnyApiFactory>;
   getFeatureFlags(): Iterable<PluginFeatureFlagConfig>;
+  getInfo(): Promise<PluginInfo>;
+  setMetadataExtender(extender: ExtendMetadata): void;
```
I'm thinking we aim to make this private API so that there's no need to consider the stability of it, what do you think? Most likely a hidden `__InternalSetMetadataExtender` is the way to go, possibly with some versioning information. My experience with using symbols or weak maps is that it just makes things more difficult to debug without much gained.

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -403,6 +405,12 @@ export type ErrorBoundaryFallbackProps = {
 };
 
+// @public
+export type ExtendMetadata = (
```
And with the above proposal we can make this private, which would be nice

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -632,14 +645,39 @@ export type PluginConfig<
   routes?: Routes;
   externalRoutes?: ExternalRoutes;
   featureFlags?: PluginFeatureFlagConfig[];
+  info?: PluginConfigInfo;
```
Same thing here, perhaps `__experimentalInfo` for now?

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -632,14 +645,39 @@ export type PluginConfig<
+  packageJson?: Record<string, unknown>;
+  description?: string;
+  version?: string;
+  ownerEntityRef?: string;
```
It may be best to allow for multiple owners tbh. Not that we would want that, but it's perhaps a bit too strict to encode that into the core APIs? Thinking just `ownerEntityRefs?: string[]`.

I very much think we should aim to have entity refs as part of the ownership, so that we can link to group pages etc. We'll need to think a bit about how to achieve that without the core API handling concerns that it shouldn't. We can lean heavily on `app-defaults` to achieve that btw. It's fine for the `app-defaults` package to have pretty broad dependencies and for example set up default route bindings or something like that.

grantila replied:
Agreed, changed to an array (although it can be configured to a single string).

We don't need to depend on anything catalog-wise, it's just that the name ownerEntityRefs is clear it its semantics.

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -632,14 +645,39 @@ export type PluginConfig<
+  version?: string;
+  ownerEntityRef?: string;
+  links: PluginInfoLink[];
+  role?: string;
```
Would this ever be anything else than `frontend-plugin`?

grantila replied:
For now, perhaps not. Although perhaps a _common_ plugin.

Rugvip replied:
Common plugins don't exist though, it's just frontend of backend, the rest are libraries that don't export anything that the framework is aware of. That's the current state at least x)

grantila replied:
Ok, removed.

Rugvip commented on packages/core-app-api/src/app/types.ts:
```diff
@@ -213,6 +214,20 @@ export type AppOptions = {
+   * The package name (key) can contain wildcards and will be matched using
+   * minimatch.
+   */
+  pluginOwners?: Record<string, string>[];
+
+  /**
+   * Decorate the PluginInfo part of a plugin
+   */
+  pluginInfoDecorator?: (plugin: PluginInfo, id: string) => void;
```
I'm not entirely sure about the split of these tbh. Could it perhaps be achieved with a single option that can be applied in different ways? Or perhaps at least group them up a bit, thinking something like:

```ts
createApp({
  pluginInfo: {
    owners: ...,
    decorator: ...,
  }
})
```

grantila replied:
Certainly. Not if the two older should be `_experimental*` prefixed though, let me push a new version of this, as you mention here - it makes sense.

Rugvip replied:
Not seeing any change here? 😅

grantila replied:
Fixed

Rugvip commented on packages/core-plugin-api/src/plugin/plugin-metadata.ts:
```diff
@@ -0,0 +1,122 @@
+      if (url.protocol !== 'http:' && url.protocol !== 'https:') {
+        url.protocol = 'https:';
+      }
+      info.links.push({
+        title: 'Package repository',
+        url: url.toString(),
+      });
```
Was gonna say that this might be a bit too opinionated to have as a link, but I guess you could override it using the decorator. Sidenote though, perhaps we should add in an icon here as well? Basically just matching [EntityLink](https://github.com/backstage/backstage/blob/a17aef1244b8d85616f18ac0ddb8699ef9d336e8/packages/catalog-model/src/entity/Entity.ts#L198) somewhat, but maybe without the type. It makes the links render a bit nicer

grantila replied:
Yeah plugin should really have icons, but not sure how that'd work. In EntityLink, it's documented as _"An optional semantic key that represents a visual icon."_. What the actual semantics are, is not very clear. It might make more sense to load that separately into the metadata, next to dyn-importing package.json, also dyn-importing an icon and add to the metadata?

Rugvip commented on packages/core-app-api/src/app/types.ts:
```diff
@@ -213,6 +214,20 @@ export type AppOptions = {
+   * The package name (key) can contain wildcards and will be matched using
+   * minimatch.
+   */
+  pluginOwners?: Record<string, string>[];
```
I'm feeling that a use-case that is a bit tricky to implement right now is to link to internal slack channels or equivalent for the different teams. Perhaps the value here should be a bit more elaborate than just the owner entity refs? That does tie into the discussion just below about different ways to structure this API surface though. Another option in mind is stuff like this:

```ts
createApp({
  pluginDecorators: [
    PluginDecorators.byPluginIds({
      ids: ['catalog', 'scaffolder'],
      info: {
        owner: 'infrared', // potential shorthand to avoid full entity ref here
        links: [{ title: 'Slack', url: '...' }]
      },
    })
  ]
})
```

You're basically left with only the info decorator, but provide some utilities to make it easier to implement common use-cases. That kind of surface I think tends to be a bit easier to evolve.

grantila replied:
I find the lack of link between a catalog group entity and its internal slack channel (or equivalent) is a bigger problem, and one that shouldn't be fixed here specifically.

If there is a way to decorate (in a processor in the catalog, or _generically_ in the frontend app) communication links to groups, then we could piggy-back on that here. There are certainly more places where that would be very useful. The group entity pages themselves, e.g.

grantila replied:
The last commit adds group links on the BackstageApp, generic to the whole app. Configurable easily from e.g. a yaml file, like plugin ownership metadata.

Rugvip reviewed (changes requested):
You know where we're at, but sorry that this took ages 😅

Bunch of discussions started up here and there. But overall feels solid and if we move it all over to experimental APIs I feel we could ship this pretty fast.

grantila commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -167,6 +163,7 @@ class AppContextImpl implements AppContext {
 export class AppManager implements BackstageApp {
   private apiHolder?: ApiHolder;
   private configApi?: ConfigApi;
+  private pluginMetadataExtender: PluginMetadataExtender;
```
Yes, indeed.

grantila commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -237,7 +240,10 @@ export class AppManager implements BackstageApp {
         //               collection and then make sure we initialize things afterwards.
-        result.collectedPlugins.forEach(plugin => this.plugins.add(plugin));
+        result.collectedPlugins.forEach(plugin => {
+          plugin.setMetadataExtender?.(this.pluginMetadataExtender.extend);
```
That's fine, the `?.(` will make it ok to be undefined. In that case, the returned value from the plugin's getInfo() will simply not include metadata.

Rugvip replied:
Yep I missed the `?.` 👍

grantila commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -181,9 +178,15 @@ export class AppManager implements BackstageApp {
     this.icons = options.icons;
-    this.plugins = new Set((options.plugins as CompatiblePlugin[]) ?? []);
+    this.plugins = new Set(
+      ((options.plugins as CompatiblePlugin[]) ?? []).map(plugin => {
+        plugin.setMetadataExtender?.(this.pluginMetadataExtender.extend);
+        return plugin;
+      }),
```
It's complicated to get away from. As said, we need to allow `setMetadataExtender` to not exist in older version of plugins, and plugins need to access metadata when doing `getInfo()`, but the metadata is injected into the _app_.

So, metadata needs to be provided from the app to the plugins, but the metadata retrieval functions are async and best invoked only when needed (on `getInfo()` calls) to prevent unnecessary dyn-loading of metadata chunks.

Hence this optional _"call this function if you need to get metadata"_ setter...

Rugvip replied:
Tbh I was kinda just thinking that we should split the iteration out into a separate `.forEach`, with the plugins assignment being separate 😅

grantila commented on packages/core-app-api/src/app/PluginMetadataDecoration.ts:
```diff
@@ -0,0 +1,63 @@
+    this.matchers = pluginOwners.flatMap(rec =>
+      Object.entries(rec).map(
+        ([pkgNamePattern, owner]) =>
+          [new Minimatch(pkgNamePattern), owner] as const,
```
WRT API stability, that's a valid concern... And yes, regex should suffice, it's up to the integrators to provide this map to begin with - it's not injected from pulled-in dependencies. RegExp is very reasonable.

grantila commented on packages/core-app-api/src/app/PluginMetadataDecoration.ts:
```diff
@@ -0,0 +1,63 @@
+
+    if (pkgName && !info.ownerEntityRef) {
+      info.ownerEntityRef =
+        this.matchOwner(pkgName) ?? pkgJson.backstage?.owner;
```
In the example:

```yaml
# plugin-metadata.yaml
owners:
  - '*techdocs*': the-doc-team
  - '@backstage/plugin-ci*': cicd-team
  - '@backstage/plugin-jenkins*': cicd-team
  - '*': the-core-team # Fallback unless specified above
```

This is what a Backstage implementation would have to match the plugins with internal corporate owners. Here, and in most cases, people working on Backstage and various plugins, rarely know what the "id" string of a plugin is. You'd have to read its source code, which is especially complicated when you pull in external plugins. A plugin _name_ is usually seen as its npm package name.

Rugvip replied:
Hmm, perhaps, although the plugin ID is the more central piece that constant across frontend/backend and what we're using in analytics, etc. I feel perhaps we should rather find ways to surface the plugin ID better if that's an issue?

By using the package name I'm a bit worried there will be confusion around library and module packages, which aren't really part of this.

grantila commented:
Updated according to comments

Rugvip commented on packages/core-app-api/api-report.md:
```diff
@@ -206,6 +207,8 @@ export type AppOptions = {
       >;
     }
   >;
+  pluginOwners?: Record<string, string>[];
+  pluginInfoDecorator?: (plugin: PluginInfo, id: string) => void;
```
Unresolved this and the one below, because I don't see them handled?

grantila replied:
Sorry about this, fixed now.

Rugvip commented on packages/core-plugin-api/api-report.md:
```diff
@@ -505,6 +513,11 @@ export type IdentityApi = {
 export const identityApiRef: ApiRef<IdentityApi>;
 
+// @public
+export type LazyLoadedPackageJson = () => Promise<{
+  default: Record<string, unknown>;
+}>;
```
Probably best avoid this alias for now, while shipping things as experimental. i.e. inline instead

grantila replied:
Removed export, and named it experimental

grantila replied:
It's used in two places (which isn't much, but more than 1), so inlining feels so-so

Rugvip reviewed (changes requested):
Another round! 🍻 

I'm a bit confused by the state of this tbh, you have a couple of comments saying stuff was pushed, but I don't see any changes? 😅 Could you update the API reports too as well btw? I don't think they're in sync atm, which makes it a lot trickier to review.

joshjung commented:
This seems like such a great idea, it's been over 2 years since this was originally opened, are there plans to finish this?

schultzp2020 commented:
@grantila this looks like a great idea! Is there anything I can do to help move this PR along?

Rugvip commented:
@schultzp2020 I'm thinking a good path forward might be to reimplement this into the new frontend system, then potentially backport if we're happy with the implementation there. I'd also think we want to bring in support for `catalog-info.yaml` or connecting to entities in the catalog in some way as well.

I think the design should be similar to that in this PR - a standardized set of metadata fields for all plugins that by default can be extracted from `package.json`/`catalog-info.yaml` but potentially also be populated manually. It might also be possible to override this metadata, either via `plugin.withOverrides` in the new frontend system and/or via config or some other means, so that 3rd party plugin support can be routed to appropriate internal teams. Lastly If we're loading in any JSON or YAML, we'll want to keep that lazy as it can get pretty large for the total number of plugins, and we wouldn't want to load that all upfront.
